pkg/mimirpb/custom.go

```diff
@@ -8,8 +8,76 @@ import (
 	"math"
 
 	"github.com/prometheus/prometheus/model/histogram"
+	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/encoding/proto"
+	"google.golang.org/grpc/mem"
+	protobufproto "google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/protoadapt"
 )
 
+func init() {
+	c := encoding.GetCodecV2(proto.Name)
+	encoding.RegisterCodecV2(&CodecV2{c})
+}
+
+// CodecV2 customizes gRPC unmarshalling.
+type CodecV2 struct {
+	encoding.CodecV2
+}
+
+func messageV2Of(v any) protobufproto.Message {
+	switch v := v.(type) {
+	case protoadapt.MessageV1:
+		return protoadapt.MessageV2Of(v)
+	case protoadapt.MessageV2:
+		return v
+	default:
+		panic(fmt.Errorf("unrecognized message type %T", v))
+	}
+}
+
+// Unmarshal customizes gRPC unmarshalling.
+// If v implements the UnmarshalerV2 interface, its SetBuffer method is called with the unmarshalling buffer.
+func (c *CodecV2) Unmarshal(data mem.BufferSlice, v any) error {
+	vv := messageV2Of(v)
+	buf := data.MaterializeToBuffer(mem.DefaultBufferPool())
+	defer buf.Free()
```

Comment:
[nit] We decrease the buffer ref counter but then later we retain it. However, before retaining it we call `Ref()` again, so I think this code is correct (this defer release it in case we `return err` below) but it's a bit tricky to understand so I would add a comment to explain.

Reply:
Yep, when the deferred call executes, the reference count will have been incremented unless an error occurred.

Reply:
Done.